src/lib/storage/table.cpp

```diff
@@ -353,36 +353,99 @@ const TableKeyConstraints& Table::soft_key_constraints() const {
+  Assert(_type == TableType::Data, "TableKeyConstraints are not tracked for reference tables across the PQP.");
 
-  // Check validity of specified columns
+  // Check validity of specified columns.
+  const auto column_count = this->column_count();
   for (const auto& column_id : table_key_constraint.columns()) {
-    Assert(column_id < column_count(), "ColumnID out of range");
+    Assert(column_id < column_count, "ColumnID out of range.");
 
-    // PRIMARY KEY requires non-nullable columns
+    // PRIMARY KEY requires non-nullable columns.
     if (table_key_constraint.key_type() == KeyConstraintType::PRIMARY_KEY) {
       Assert(!column_is_nullable(column_id), "Column must be non-nullable to comply with PRIMARY KEY.");
     }
   }
 
   {
-    auto scoped_lock = acquire_append_mutex();
+    const auto scoped_lock = acquire_append_mutex();
 
     for (const auto& existing_constraint : _table_key_constraints) {
       // Ensure that no other PRIMARY KEY is defined
       Assert(existing_constraint.key_type() == KeyConstraintType::UNIQUE ||
                  table_key_constraint.key_type() == KeyConstraintType::UNIQUE,
-             "Another primary key already exists for this table.");
+             "Another primary TableKeyConstraint exists for this table.");
 
       // Ensure there is only one key constraint per column set.
       Assert(table_key_constraint.columns() != existing_constraint.columns(),
-             "Another key constraint for the same column set has already been defined.");
+             "Another TableKeyConstraint for the same column set has already been defined.");
     }
 
     _table_key_constraints.insert(table_key_constraint);
   }
 }
 
+void Table::add_soft_foreign_key_constraint(const ForeignKeyConstraint& foreign_key_constraint) {
+  Assert(_type == TableType::Data, "ForeignKeyConstraints are not tracked for reference tables across the PQP.");
+
+  Assert(&*foreign_key_constraint.foreign_key_table() == &*this, "ForeignKeyConstraint is added to the wrong table.");
+
+  // Check validity of specified columns.
+  const auto column_count = this->column_count();
+  for (const auto& column_id : foreign_key_constraint.foreign_key_columns()) {
+    Assert(column_id < column_count, "ColumnID out of range.");
+  }
+
+  const auto referenced_table = foreign_key_constraint.table();
+  Assert(referenced_table, "ForeignKeyConstraint must reference an existing table.");
+
+  // Check validity of key columns from other table.
+  const auto referenced_table_column_count = referenced_table->column_count();
+  for (const auto& column_id : foreign_key_constraint.columns()) {
+    Assert(column_id < referenced_table_column_count, "ColumnID out of range.");
+  }
+
+  {
+    const auto scoped_lock = acquire_append_mutex();
+    Assert(!_foreign_key_constraints.contains(foreign_key_constraint), "ForeignKeyConstraint is already set.");
```

Comment:
This will not be measurable, but: you could first insert and then check the result for the bool value. This way, we have only one lookup into the set.

Otherwise, over the years the caused CO2 by this inefficiency can quickly sum up to a flight between Berlin and San Fransisco.

Reply:
You're absolutely right. We could as well do nothing. Inserting the exact same constraint into a set is not a problem and just a leftover from when we used vectors to store constraints. However, if I think about it, you maybe should not be able to add FK constraints that have a subset/superset of columns (e.g., when having [a, b] in [c, d], you should not add [a] in [c] or [b] in [d] and vice versa) - only the "largest" constraint should be added then, right?

Reply:
I have no idea what you are talking about.